File: 백준/Silver/15654. N과 M （5）/N과 M （5）.js
const [[N, M], arr] = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n').map(line => line.trim().split(' ').map(Number));
arr.sort((a, b) => a - b);
const visited = new Array(N).fill(false);
const tmp = new Array(M);
const ans = [];
DFS(0);
console.log(ans.map(line => line.join(' ')).join('\n'));
function DFS(depth) {
  if(depth === M) {
    ans.push([...tmp]);
    return;
  }
  for(let i = 0; i < N; i++) {
    if(visited[i]) continue;
    visited[i] = true;
    tmp[depth] = arr[i];
    DFS(depth + 1);
    visited[i] = false;
  }
}

File: 백준/Silver/15654. N과 M （5）/N과 M （5）.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, M;
    static int[] arr, tmp;
    static boolean[] visited;
    static StringBuilder sb = new StringBuilder();
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        arr = new int[N];
        tmp = new int[M];
        visited = new boolean[N];
        st = new StringTokenizer(br.readLine());
        for(int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(arr);
        DFS(0);
        System.out.print(sb);
    }

    static void DFS(int depth) {
        if(depth == M) {
            for(int i = 0; i < M; i++) {
                sb.append(tmp[i]);
                if(i != M - 1) sb.append(" ");
            }
            sb.append("\n");
            return;
        }
        for(int i = 0; i < N; i++) {
            if(visited[i]) continue;
            visited[i] = true;
            tmp[depth] = arr[i];
            DFS(depth + 1);
            visited[i] = false;
        }
    }
}
